appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/visitor/OpenApiWalker.java

```diff
@@ -153,113 +179,141 @@ public void accept(ApiVisitor visitor) {
+        processAnnotations(SecurityRequirement.class, visitor::visitSecurityRequirement, SecurityRequirements.class);
+        processAnnotations(SecurityRequirements.class, visitor::visitSecurityRequirements, SecurityRequirement.class);
 
         // JAX-RS response types
         processAnnotations(Produces.class, visitor::visitProduces);
         processAnnotations(Consumes.class, visitor::visitConsumes);
+
+        // OpenAPI response types
         processAnnotations(RequestBody.class, visitor::visitRequestBody);
     }
 
-    private <A extends Annotation, A2 extends Annotation, E extends AnnotatedElement> void processAnnotations(
-            Class<A> annotationClass, VisitorFunction<A, E> annotationFunction, Class<A2> altClass,
-            VisitorFunction<A2, E> altFunction, boolean allowInterfaces) {
+    @SafeVarargs
+    private final <A extends Annotation, E extends AnnotatedElement> void processAnnotations(
+            Class<A> annotationClass, VisitorFunction<A, E> annotationFunction, 
+            Class<? extends Annotation>... alternatives) {
         for (Class<?> clazz : classes) {
+            processAnnotation(clazz, annotationClass, annotationFunction, alternatives);
+        }
+    }
 
-            // Don't read interfaces if the annotation should only be read from concrete classes
-            if (!allowInterfaces && (Modifier.isAbstract(clazz.getModifiers()) || clazz.isInterface())) {
-                continue;
-            }
-
-            processAnnotation(clazz, annotationClass, annotationFunction, altClass, altFunction,
-                    new OpenApiContext(api, getResourcePath(clazz, resourceMapping)));
+    @SafeVarargs
+    private final <T, A extends Annotation, E extends AnnotatedElement> void processAnnotation(
+            Class<T> annotatedClass, Class<A> annotationClass, VisitorFunction<A, E> annotationFunction,
+            Class<? extends Annotation>... alternatives) {
+        AnnotationInfo<T> annotations = AnnotationInfo.valueOf(annotatedClass);
+        processAnnotation(annotatedClass, annotationClass, annotationFunction, annotations,
+                new OpenApiContext(classes, api, getResourcePath(annotatedClass, resourceMapping)), alternatives);
 
-            for (Field field : clazz.getDeclaredFields()) {
-                processAnnotation(field, annotationClass, annotationFunction, altClass, altFunction,
-                        new OpenApiContext(api, null));
+        for (Field field : annotatedClass.getDeclaredFields()) {
+            if (annotations.isAnnotationPresent(annotationClass, field)) {
+                if (   annotationClass == HeaderParam.class
+                    || annotationClass == CookieParam.class
+                    || annotationClass == PathParam.class
+                    || annotationClass == QueryParam.class) {
+                    // NB. if fields are annotated as Param all methods have it
```

Comment:
FYI: The field => method logic part was moved here from the application processor